lib/node_modules/@stdlib/repl/lib/pager.js

```diff
@@ -0,0 +1,217 @@
+
+'use strict';
+
+// MODULES //
+
+var readline = require( 'readline' );
+var logger = require( 'debug' );
+var setNonEnumerableReadOnly = require( '@stdlib/utils/define-nonenumerable-read-only-property' );
+var max = require( '@stdlib/math/base/special/max' );
+var repeat = require( '@stdlib/string/repeat' );
+var displayPrompt = require( './display_prompt.js' );
+
+
+// VARIABLES //
+
+var debug = logger( 'repl:pager' );
+
+
+// MAIN //
+
+/**
+* Constructor for creating a pager.
+*
+* @private
+* @constructor
+* @param {REPL} repl - REPL instance
+* @param {WritableStream} ostream - writable stream
+* @param {Function} ostreamWrite - default ostream `write`
+* @returns {Pager} pager instance
+*/
+function Pager( repl, ostream, ostreamWrite ) {
+	if ( !(this instanceof Pager) ) {
+		return new Pager( repl, ostream, ostreamWrite );
+	}
+	debug( 'Creating a pager...' );
+
+	// Cache a reference to the provided REPL instance:
+	this.repl = repl;
+
+	// Cache a reference to the output writable stream:
+	this._ostream = ostream;
+
+	// Cache a reference to the ostream write callback:
+	this._write = ostreamWrite;
+
+	// Initialize a buffer to store the output viewport height:
+	this._viewportHeight = -1;
+
+	// Initialize a buffer to store the index of the current scroll position:
+	this._startIndex = -1;
+
+	// Initialize a buffer to store entire output data:
+	this._data = '';
+
+	// Initialize a buffer containing all lines of the output:
+	this._lines = [];
+
+	return this;
+}
+
+/**
+* Displays the scroll view of the initial page.
+*
+* @name _displayPage
+* @memberof Pager.prototype
+* @returns {void}
+*/
+setNonEnumerableReadOnly( Pager.prototype, '_displayPage', function displayPage() {
+	var i;
+
+	// Display the initial page till the `viewportHeight`:
+	this._write.call( this._ostream, '\n' );
+	for ( i = 0; i < this._viewportHeight; i++ ) {
+		this._write.call( this._ostream, this._lines[ i ] + '\n' );
+	}
+	// Draw a seperator to denote that the output is scrollable below:
+	this._write.call( this._ostream, repeat( '_', this._ostream.columns ) + '\n' );
+
+	// Display scroll instructions:
+	this._write.call( this._ostream, '\u001b[1mUse UP/DOWN arrow keys to scroll. Press CTRL+X to exit...\u001b[22m' );
+	readline.moveCursor( this._ostream, 0, -1 * ( this._viewportHeight + 3 ) );
+});
+
+/**
+* Scrolls the current page based on the updated index.
+*
+* @name _scroll
+* @memberof Pager.prototype
+* @returns {void}
+*/
+setNonEnumerableReadOnly( Pager.prototype, '_scroll', function scroll() {
+	var i;
+
+	// Clear previous output:
+	readline.clearScreenDown( this._ostream );
+
+	// If scroll not at the start of the page, draw a seperator to denote that the output is scrollable above:
+	if ( this._startIndex === 0 ) {
+		this._write.call( this._ostream, '\n\n' );
+	} else {
+		this._write.call( this._ostream, '\n' + repeat( '_', this._ostream.columns ) + '\n' );
```

Comment:
Are `columns` and `rows` public APIs? You're also making an assumption that the output stream is `stdout`/`stderr`.

Reply:
I see `columns` and `rows` are coming from TTY: https://nodejs.org/api/tty.html.

Reply:
The output stream is set to the stream we set as the `output` when creating the readline interface. Shouldn't it be that way or am I missing something?

Reply:
The question concerns whether `columns` and `rows` are always available. The short answer is no, they are not. If I set the `output` option to a non-TTY stream, there are no such properties, as, in general, WriteableStreams don't have any such need.

Reply:
Meaning, in order to figure out the viewport height and width, you are assuming TTY streams. But this returns back to the question as to what to do in non-TTY contexts, and the answer to that is that we should likely not support a pager.

Reply:
@kgryte Understood and agreed👍, the pager command can simply print the output normally in non-TTY contexts...